### PR TITLE
openapi schema subcomponents, components with generics, generic subcomponents...

### PR DESCRIPTION
**Pull request by E-gy**
TLDR: This allows usage of `schema(component)` on parametrized types.
```rust
#[schema(component = "GenericStruct")]
struct GenericStruct<A, B> {
    a: A,
    b: B,
}
```
Additionally: This fixes subcomponents-appearing-only-in-other-components not appearing in the final spec 🐛.

Currently it's simply impossible due to name conflict.

Usage of generic components requires that all types be themselves components or primitives (or arrays of).

It works by appending parametrization suffix to component's name. The suffix is assembled from component names of the generic parameters (and values for generic constants). For OpenAPI primitives, it's their shorthand names. Arrays get a special pre/suffix around inner type.
~~All of this means that as long as all types used in generic params are themselves components (or primitives), the generated suffix is consistent.~~ Basically this suffix is consistent, since anonymous/inline parameters fail to generate components (and every component's name is consistently defined, recursively now).

### CODE REVIEW

**Comment by E-gy** [comment]
Just noticed this conflicts #59.
#59 should be merged first, probably.
I'll resolve conflicts either way.

**Comment by E-gy** [comment]
Now this PR also fixes a bug (reminiscent of #38) where components exclusively nested in other components weren't included in final spec.

**Comment by E-gy** [comment]
This implied fixing impl of `Schema` for `Option<T>` and `[T]` to
- themselves be components when `T` is a component
- copy-extend only `T`'s schema (instead of every sub-component).

**Comment by E-gy** [comment]
Ironically, atm generic nested subcomponents don't quite work :D
